pkg/storagev2/leveldb/store.go

```diff
@@ -0,0 +1,292 @@
+	keys := util.BytesPrefix([]byte(key.Namespace() + "/"))
+
+	iter := s.DB.NewIterator(keys, nil)
+	defer iter.Release()
+
+	for iter.Next() {
+		c++
+	}
+
+	return c, iter.Error()
+}
+
+func (s *Store) Put(item storageV2.Item) error {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	value, err := item.Marshal()
+	if err != nil {
+		return fmt.Errorf("failed serializing %w", err)
+	}
+
+	key := []byte(item.Namespace() + "/" + item.ID())
+
+	return s.DB.Put(key, value, &opt.WriteOptions{Sync: true})
+}
+
+func (s *Store) Get(item storageV2.Item) error {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(item.Namespace() + "/" + item.ID())
+
+	val, err := s.DB.Get(key, nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return storageV2.ErrNotFound
+		}
+		return err
+	}
+
+	if err = item.Unmarshal(val); err != nil {
+		return fmt.Errorf("failed decoding value %w", err)
+	}
+
+	return nil
+}
+
+func (s *Store) Has(sKey storageV2.Key) (exists bool, err error) {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(sKey.Namespace() + "/" + sKey.ID())
+
+	return s.DB.Has(key, nil)
+}
+
+func (s *Store) GetSize(sKey storageV2.Key) (size int, err error) {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(sKey.Namespace() + "/" + sKey.ID())
+
+	val, err := s.DB.Get(key, nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return 0, storageV2.ErrNotFound
+		}
+		return 0, err
+	}
+
+	return len(val), nil
+}
+
+func (s *Store) Delete(item storageV2.Key) (err error) {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(item.Namespace() + "/" + item.ID())
+
+	return s.DB.Delete(key, &opt.WriteOptions{Sync: true})
+}
+
+func (s *Store) Iterate(q storageV2.Query, fn storageV2.IterateFn) error {
+	if err := q.Validate(); err != nil {
+		return fmt.Errorf("failed iteration: %w", err)
+	}
+
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	var retErr *multierror.Error
+
+	namespace := q.Factory().Namespace()
+	prefix := namespace + "/"
+
+	keys := util.BytesPrefix([]byte(prefix))
+
+	iter := s.DB.NewIterator(keys, nil)
+
+	switch q.Order {
+	case storageV2.KeyAscendingOrder:
+		for iter.Next() {
+			k := string(iter.Key())
+			v := iter.Value()
+
+			stop, err := apply(k, v, q, fn)
+			if err != nil {
+				retErr = multierror.Append(retErr, err)
+			}
+			if stop {
+				break
+			}
+		}
+	case storageV2.KeyDescendingOrder:
+		for ok := iter.Last(); ok; ok = iter.Prev() {
+			k := string(iter.Key())
+			v := iter.Value()
+
+			stop, err := apply(k, v, q, fn)
+			if err != nil {
+				retErr = multierror.Append(retErr, err)
+			}
+			if stop {
+				break
+			}
+		}
+	}
+
+	iter.Release()
+
+	if err := iter.Error(); err != nil {
+		retErr = multierror.Append(retErr, err)
+	}
+
+	return retErr.ErrorOrNil()
+}
+
+func apply(k string, v []byte, q storageV2.Query, fn func(storageV2.Result) (bool, error)) (stop bool, err error) {
+	namespace := q.Factory().Namespace()
```

Comment:
So this can be avoided. Basically here for each iteration we are allocating 2 objects. The namespace is constant, so we dont need to use Factory here.

Reply:
we're allocating on the stack, since the var does not escape func scope, so no pressure on gc.
but yes, I can pass it in as a value